core/src/main/java/io/camunda/migrator/RuntimeMigrator.java

```diff
@@ -42,101 +43,139 @@ public class RuntimeMigrator {
   @Autowired
   protected CamundaClient camundaClient;
 
-  public void migrate() {
-    String latestLegacyId = idKeyMapper.findLatestIdByType("runtimeProcessInstance");
-    ProcessInstanceQuery processInstanceQuery = ((ProcessInstanceQueryImpl) runtimeService.createProcessInstanceQuery())
-        .idAfter(latestLegacyId)
-        .rootProcessInstances();
-
-    long maxLegacyProcessInstanceCount = processInstanceQuery.count();
-    for (int i = 0; i < maxLegacyProcessInstanceCount; i = i + BATCH_SIZE - 1) {
-      processInstanceQuery.listPage(i, BATCH_SIZE).forEach(legacyProcessInstance -> {
-        String legacyId = legacyProcessInstance.getId();
-        var idKeyDbModel = new IdKeyDbModel();
-        idKeyDbModel.setId(legacyId);
-        idKeyDbModel.setKey(null);
-        idKeyDbModel.setType("runtimeProcessInstance");
-        idKeyMapper.insert(idKeyDbModel);
-      });
-    }
+  protected boolean retryMode = false;
 
-    // TODO: paginate this
-    idKeyMapper.findProcessInstanceIds().forEach(legacyProcessInstanceId -> {
-      Map<String, Object> globalVariables = new HashMap<>();
+  public void migrate() {
+    List<String> processInstanceIds = fetchProcessInstancesToMigrate();
+
+    processInstanceIds.forEach(legacyProcessInstanceId -> {
+      if (validateProcessInstanceMigration(legacyProcessInstanceId)) {
+        long processInstanceKey = startNewProcessInstance(legacyProcessInstanceId);
+        insertRuntimeProcessInstanceEntity(legacyProcessInstanceId, processInstanceKey);
+      } else {
+        System.out.println("Skipping process instance with legacyId " + legacyProcessInstanceId); // TODO log
+        insertRuntimeProcessInstanceEntity(legacyProcessInstanceId, null);
+      }
+    });
 
-      runtimeService.createVariableInstanceQuery()
-          .activityInstanceIdIn(legacyProcessInstanceId)
-          .list()
-          .forEach(variable -> globalVariables.put(variable.getName(), variable.getValue())); // Collectors#toMap cannot handle null values and throws NPE.
+    activateMigratorJobs();
+  }
 
-      globalVariables.put("legacyId", legacyProcessInstanceId);
+  private List<String> fetchProcessInstancesToMigrate() {
```

Comment:
```suggestion
  private Set<String> fetchProcessInstancesToMigrate() {
```
Can this be a `Set`?